demo/runtime/src/test_helper_pallet.rs

```diff
@@ -67,14 +76,27 @@ pub mod pallet {
 		}
 	}
 
-	impl<T: Config> pallet_address_associations::OnNewAssociation<AccountId> for Pallet<T> {
+	impl<T: Config>
+		pallet_address_associations::OnNewAssociation<
+			AccountId,
+			<T as frame_system::Config>::AccountId,
+		> for Pallet<T>
+	{
 		fn on_new_association(
+			origin_account: <T as frame_system::Config>::AccountId,
 			partner_chain_address: AccountId,
 			main_chain_key_hash: MainchainKeyHash,
-		) {
+		) -> bool {
 			log::info!(
 				"New address association: {partner_chain_address:?} -> {main_chain_key_hash:?}"
 			);
+			T::Currency::withdraw(
+				&origin_account,
+				T::XAssociationBurnAmount::get(),
+				WithdrawReasons::FEE,
+				ExistenceRequirement::KeepAlive,
+			)
+			.is_ok()
```

Comment:
This time the logic of incurring additional fee is not in the address-association pallet but in the user runtime (in the test-helper-pallet in this instance)